pretrain/shakespeare.py

```diff
@@ -70,7 +70,7 @@ def main() -> None:
 
     model = fabric.setup_module(model)
 
-    optimizer = torch.optim.AdamW(model.parameters(), lr=learning_rate, weight_decay=weight_decay, betas=(beta1, beta2))
+    optimizer = torch.optim.AdamW(model.parameters(), lr=learning_rate, weight_decay=weight_decay, betas=(beta1, beta2), foreach=False)
```

Comment:
We have several other uses of optimizers in other files. Should they be updated too?

From the PR description, is the simple rule that if you are limited by memory, you want to set `foreach=False`?

Do you think it would be a good idea to come up with some heuristic for this based on the model size? In Parrot we support sizes from 70m to 40b so setting this could be detrimental to the performance of smaller models.